Python 3.8/gettext: replace custom pgettext implementation with built--in pgettext

josephsl opened the pull request:
Hi,

This will impact localization and resolves a long-standing issue with it:

### Link to issue number:
Closes #9207

### Summary of the issue:
Python 3.8 introduces gettext.pgettext to obtain translated text based on context. Prior to this, NVDA defined a custom implementation of pgettext function.

### Description of how this pull request fixes the issue:
Replace custom pgettext implementation and built-in namespace assignment with Python 3.8's gettext.pgettext by using translation.install function's "names" sequence parameter.

### Testing strategy:
Two ways:

1. Perform a diff between current master and this PR's locale/* data after generating pot files.
2. Manual: have translators test this change to make sure localizations are correct.

### Known issues with pull request:
None

### Change log entry:
None needed, unless a note to developers is needed to highlight pgettext function.

### Code Review Checklist:

This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review and confirm you have considered the following items.
Mark items you have considered by checking them.
You can do this when editing the Pull request description with an x: `[ ]` becomes `[x]`.
You can also check the checkboxes after the PR is created.

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual tests.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.

josephsl committed:
Add-on handler/pgettext: remove pgettext namespace definition. Re #9207.

As gettext.pgettext is present, there is no need to manually assign NVDA's own pgettext function to built-in namespace for add-ons. Therefore remove this assignment.

josephsl committed:
Language handler/pgettext: remove nVDA-specific pgettext function. re #9207.

As Python 3.8 includes gettext.pgettext, there is no need for NVDA's implementation of pgettext function. Therefore remove makePgettext and builtins namespace assignment from language handler.

josephsl committed:
Language handler: remove unused builtins namespace import.

josephsl committed:
Language handler/pgettext: add 'pgettext' to built-in namespace. Re #9207.

As Python 3.8 includes gettext.pgettext, add it to built-in namespace by using 'names' parameter of translation.install method.

josephsl committed:
Update copyright header

josephsl commented:
Hi,

Note that this change may affect add-ons. As far as I know, all add-ons hosted on community add-ons website are not affected.

Thanks.

CyrilleB79 commented:
Hi

You write:
> Note that this change may affect add-ons.

In this case, a point should be added in the changes for developpers describing what is the impact on add-ons. Thanks.

josephsl commented:
Hi, the impact has to do with localizations, no changes to function calls and such. Thanks.

lukaszgo1 commented:
Have you tested that with this change `pgettext`` works correctly when used for translatable strings in add-ons?

josephsl committed:
Revert "Add-on handler/pgettext: remove pgettext namespace definition. Re #9207."

This reverts commit 991de433b645a2a45c7319a9ca2e88c8813330a8 so add-on translations will work.

josephsl commented:
Hi, yes. The needed change is assigning Python’s pgettext function to add-on namespace (done in the upcoming commit). To answer another question, no it will not require changes to add-ons at all. Thanks.

josephsl committed:
Add-on handler/pgettext: replace NVDA-specific pgettext call with Python's pgettext function. Re #9207.

Add-on initialization will still require assigning pgettext to add-on namespace, therefore use translations.pgettext instead of now removed NVDA-specific implementation.
